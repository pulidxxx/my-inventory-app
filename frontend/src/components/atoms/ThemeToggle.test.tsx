import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ThemeToggle } from "./ThemeToggle";

// Mock de localStorage
const mockLocalStorage = (() => {
    let store: Record<string, string> = {};

    return {
        getItem: (key: string) => store[key] || null,
        setItem: (key: string, value: string) => {
            store[key] = value.toString();
        },
        clear: () => {
            store = {};
        },
    };
})();

// Mock de document.body.setAttribute
const mockSetAttribute = jest.fn();

beforeAll(() => {
    Object.defineProperty(window, "localStorage", {
        value: mockLocalStorage,
    });

    Object.defineProperty(document.body, "setAttribute", {
        value: mockSetAttribute,
    });
});

beforeEach(() => {
    // Limpiar mocks antes de cada prueba
    mockLocalStorage.clear();
    mockSetAttribute.mockClear();
});

describe("ThemeToggle Component", () => {
    it("renders correctly with light mode as default when no theme is set", () => {
        render(<ThemeToggle />);

        const toggle = screen.getByRole("checkbox");
        expect(toggle).not.toBeChecked();
        expect(screen.getByText("Modo claro")).toBeInTheDocument();
    });

    it("renders in dark mode when localStorage has 'dark' theme", () => {
        mockLocalStorage.setItem("theme", "dark");
        render(<ThemeToggle />);

        const toggle = screen.getByRole("checkbox");
        expect(toggle).toBeChecked();
        expect(screen.getByText("Modo oscuro")).toBeInTheDocument();
    });

    it("toggles between light and dark mode when clicked", async () => {
        render(<ThemeToggle />);
        const toggle = screen.getByRole("checkbox");

        // De light a dark
        await userEvent.click(toggle);
        expect(toggle).toBeChecked();
        expect(screen.getByText("Modo oscuro")).toBeInTheDocument();
        expect(mockLocalStorage.getItem("theme")).toBe("dark");
        expect(mockSetAttribute).toHaveBeenCalledWith("data-theme", "dark");

        // De dark a light
        await userEvent.click(toggle);
        expect(toggle).not.toBeChecked();
        expect(screen.getByText("Modo claro")).toBeInTheDocument();
        expect(mockLocalStorage.getItem("theme")).toBe("light");
        expect(mockSetAttribute).toHaveBeenCalledWith("data-theme", "light");
    });

    it("updates document body attribute on mount", () => {
        mockLocalStorage.setItem("theme", "dark");
        render(<ThemeToggle />);
        expect(mockSetAttribute).toHaveBeenCalledWith("data-theme", "dark");
    });

    it("initializes with light mode if localStorage has invalid value", () => {
        mockLocalStorage.setItem("theme", "invalid");
        render(<ThemeToggle />);

        const toggle = screen.getByRole("checkbox");
        expect(toggle).not.toBeChecked();
        expect(screen.getByText("Modo claro")).toBeInTheDocument();
    });

    it("matches snapshot in light mode", () => {
        const { asFragment } = render(<ThemeToggle />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot in dark mode", () => {
        mockLocalStorage.setItem("theme", "dark");
        const { asFragment } = render(<ThemeToggle />);
        expect(asFragment()).toMatchSnapshot();
    });
});
