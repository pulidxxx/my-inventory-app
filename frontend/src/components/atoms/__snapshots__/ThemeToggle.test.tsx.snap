// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggle Component matches snapshot in dark mode 1`] = `
<DocumentFragment>
  <label
    class="inline-flex items-center cursor-pointer fixed bottom-4 right-4 z-50"
  >
    <input
      checked=""
      class="sr-only peer"
      type="checkbox"
    />
    <div
      class="relative w-11 h-6 bg-sky-200 rounded-full peer peer-focus:ring-4 peer-focus:ring-blue-300 dark:peer-focus:ring-blue-800 dark:bg-gray-700 peer-checked:after:translate-x-full rtl:peer-checked:after:-translate-x-full peer-checked:after:border-white after:content-[''] after:absolute after:top-0.5 after:start-[2px] after:bg-white after:border-gray-300 after:border after:rounded-full after:h-5 after:w-5 after:transition-all dark:border-gray-600 peer-checked:bg-blue-600 dark:peer-checked:bg-blue-600"
    />
    <span
      class="ms-3 text-sm font-medium text-sky-400 dark:text-gray-300"
    >
      Modo oscuro
    </span>
  </label>
</DocumentFragment>
`;

exports[`ThemeToggle Component matches snapshot in light mode 1`] = `
<DocumentFragment>
  <label
    class="inline-flex items-center cursor-pointer fixed bottom-4 right-4 z-50"
  >
    <input
      class="sr-only peer"
      type="checkbox"
    />
    <div
      class="relative w-11 h-6 bg-sky-200 rounded-full peer peer-focus:ring-4 peer-focus:ring-blue-300 dark:peer-focus:ring-blue-800 dark:bg-gray-700 peer-checked:after:translate-x-full rtl:peer-checked:after:-translate-x-full peer-checked:after:border-white after:content-[''] after:absolute after:top-0.5 after:start-[2px] after:bg-white after:border-gray-300 after:border after:rounded-full after:h-5 after:w-5 after:transition-all dark:border-gray-600 peer-checked:bg-blue-600 dark:peer-checked:bg-blue-600"
    />
    <span
      class="ms-3 text-sm font-medium text-sky-400 dark:text-gray-300"
    >
      Modo claro
    </span>
  </label>
</DocumentFragment>
`;
